File: src/crypto.ts
import { argon2id, argon2Verify } from "hash-wasm";

const ARGON_OPTS = {
  parallelism: 1,
  iterations: 3,
  memorySize: 65536,
  hashLength: 32,
} as const;

function randomBytes(n: number): Uint8Array {
  return crypto.getRandomValues(new Uint8Array(n));
}

function toB64(buf: Uint8Array): string {
  return Buffer.from(buf).toString("base64");
}

function fromB64(s: string): Uint8Array {
  return new Uint8Array(Buffer.from(s, "base64"));
}

export async function hashMasterPassword(password: string): Promise<string> {
  const salt = randomBytes(16);
  return argon2id({
    password,
    salt,
    ...ARGON_OPTS,
    outputType: "encoded",
  });
}

export async function verifyMasterPassword(
  password: string,
  encodedHash: string,
): Promise<boolean> {
  try {
    return await argon2Verify({ password, hash: encodedHash });
  } catch {
    return false;
  }
}

/** Derive 32-byte AES key from master password + salt (Argon2id). */
export async function deriveAesKey(
  masterPassword: string,
  saltB64: string,
): Promise<CryptoKey> {
  const raw = await argon2id({
    password: masterPassword,
    salt: fromB64(saltB64),
    ...ARGON_OPTS,
    outputType: "binary",
  });
  return crypto.subtle.importKey("raw", raw as Uint8Array<ArrayBuffer>, { name: "AES-GCM" }, false, [
    "encrypt",
    "decrypt",
  ]);
}

export function newKdfSalt(): string {
  return toB64(randomBytes(16));
}

/** AES-256-GCM: returns base64(iv || ciphertext+tag). */
export async function encryptSecret(
  key: CryptoKey,
  plaintext: string,
): Promise<string> {
  const iv = randomBytes(12);
  const encoded = new TextEncoder().encode(plaintext);
  const cipher = new Uint8Array(
    await crypto.subtle.encrypt({ name: "AES-GCM", iv: iv as Uint8Array<ArrayBuffer> }, key, encoded as Uint8Array<ArrayBuffer>),
  );
  const out = new Uint8Array(iv.length + cipher.length);
  out.set(iv, 0);
  out.set(cipher, iv.length);
  return toB64(out);
}

export async function decryptSecret(
  key: CryptoKey,
  payloadB64: string,
): Promise<string> {
  const data = fromB64(payloadB64);
  const iv = data.slice(0, 12);
  const cipher = data.slice(12);
  const plain = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv },
    key,
    cipher,
  );
  return new TextDecoder().decode(plain);
}

export async function encryptBytes(
  key: CryptoKey,
  data: Uint8Array,
): Promise<Uint8Array> {
  const iv = randomBytes(12);
  const cipher = new Uint8Array(
    await crypto.subtle.encrypt({ name: "AES-GCM", iv: iv as Uint8Array<ArrayBuffer> }, key, data as Uint8Array<ArrayBuffer>),
  );
  const out = new Uint8Array(iv.length + cipher.length);
  out.set(iv, 0);
  out.set(cipher, iv.length);
  return out;
}

export async function decryptBytes(
  key: CryptoKey,
  data: Uint8Array,
): Promise<Uint8Array> {
  const iv = data.slice(0, 12);
  const cipher = data.slice(12);
  return new Uint8Array(
    await crypto.subtle.decrypt({ name: "AES-GCM", iv }, key, cipher),
  );
}

File: src/internal/crypto/crypto.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"io"

	"github.com/alexedwards/argon2id"
	"golang.org/x/crypto/argon2"
)

var argonParams = &argon2id.Params{
	Memory:      65536,
	Iterations:  3,
	Parallelism: 1,
	SaltLength:  16,
	KeyLength:   32,
}

func randomBytes(n int) ([]byte, error) {
	buf := make([]byte, n)
	if _, err := io.ReadFull(rand.Reader, buf); err != nil {
		return nil, err
	}
	return buf, nil
}

func toB64(buf []byte) string {
	return base64.StdEncoding.EncodeToString(buf)
}

func fromB64(s string) ([]byte, error) {
	return base64.StdEncoding.DecodeString(s)
}

func HashMasterPassword(password string) (string, error) {
	return argon2id.CreateHash(password, argonParams)
}

func VerifyMasterPassword(password, encodedHash string) bool {
	match, err := argon2id.ComparePasswordAndHash(password, encodedHash)
	if err != nil {
		return false
	}
	return match
}

func DeriveAESKey(masterPassword, saltB64 string) ([]byte, error) {
	salt, err := fromB64(saltB64)
	if err != nil {
		return nil, err
	}
	key := argon2.IDKey([]byte(masterPassword), salt, argonParams.Iterations, argonParams.Memory, argonParams.Parallelism, argonParams.KeyLength)
	return key, nil
}

func NewKdfSalt() (string, error) {
	salt, err := randomBytes(16)
	if err != nil {
		return "", err
	}
	return toB64(salt), nil
}

func EncryptSecret(key []byte, plaintext string) (string, error) {
	iv, err := randomBytes(12)
	if err != nil {
		return "", err
	}
	ciphertext, err := encryptGCM(key, iv, []byte(plaintext))
	if err != nil {
		return "", err
	}
	out := append(iv, ciphertext...)
	return toB64(out), nil
}

func DecryptSecret(key []byte, payloadB64 string) (string, error) {
	data, err := fromB64(payloadB64)
	if err != nil {
		return "", err
	}
	if len(data) < 12 {
		return "", fmt.Errorf("invalid secret payload")
	}
	plain, err := decryptGCM(key, data[:12], data[12:])
	if err != nil {
		return "", err
	}
	return string(plain), nil
}

func EncryptBytes(key, data []byte) ([]byte, error) {
	iv, err := randomBytes(12)
	if err != nil {
		return nil, err
	}
	ciphertext, err := encryptGCM(key, iv, data)
	if err != nil {
		return nil, err
	}
	out := make([]byte, len(iv)+len(ciphertext))
	copy(out, iv)
	copy(out[len(iv):], ciphertext)
	return out, nil
}

func DecryptBytes(key, data []byte) ([]byte, error) {
	if len(data) < 12 {
		return nil, fmt.Errorf("invalid encrypted data")
	}
	return decryptGCM(key, data[:12], data[12:])
}

func encryptGCM(key, iv, plaintext []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	return gcm.Seal(nil, iv, plaintext, nil), nil
}

func decryptGCM(key, iv, ciphertext []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	return gcm.Open(nil, iv, ciphertext, nil)
}
